File: lib/rgfa/line/edge/gfa2/other.rb
RGFA::Line::Edge::GFA2 ||= Module.new

module RGFA::Line::Edge::GFA2::Other

  # @param oriented_segment [RGFA::OrientedLine]
  #   one of the two oriented segments of the line
  # @return [RGFA::OrientedLine] the other oriented segment
  # @raise [RGFA::NotFoundError] if segment_end is not a segment end of the line
  def other_oriented_segment(oriented_segment)
    if (sid1 == oriented_segment)
      return sid2
    elsif (sid2 == oriented_segment)
      return sid1
    else
      raise RGFA::NotFoundError,
        "Oriented segment '#{oriented_segment}' not found\n"+
        "Line: #{self}"
    end
  end

  # The other segment of a connection line
  # @param segment [RGFA::Line::Segment::GFA2, Symbol] segment name or instance
  # @raise [RGFA::NotFoundError]
  #   if segment is not involved in the connection
  # @return [RGFA::Line::Segment::GFA2, Symbol] the instance or symbol
  #   of the other segment of the connection
  #   (which is the +segment+ itself, when the connection is circular)
  def other(segment)
    segment_name = segment.to_sym
    if segment_name == sid1.name
      sid2.line
    elsif segment_name == sid2.name
      sid1.line
    else
      raise RGFA::NotFoundError,
        "Line #{self} does not involve segment #{segment_name}"
    end
  end

end

File: gfapy/line/edge/gfa2/other.py
import gfapy

class Other:

  def other_oriented_segment(self, oriented_segment):
    """The other oriented segment.

    Parameters:
      oriented_segment (gfapy.OrientedLine) : One of the two oriented segments of the line.

    Returns:
      gfapy.OrientedLine

    Raises:
      gfapy.error.NotFoundError: If segment_end is not a segment end of the line.
    """
    if (self.sid1 == oriented_segment):
      return self.sid2
    elif (self.sid2 == oriented_segment):
      return self.sid1
    else:
      raise gfapy.NotFoundError(
          "Oriented segment '{}' not found\n".format(oriented_segment) +
          "Line: {}".format(self))

  def other(self, segment):
    """The other segment of an edge line.

    Parameters:
      segment (gfapy.line.segment.GFA2) : Segment name or instance.

    Raises:
      gfapy.error.NotFoundError: If segment is not a segment of the line.

    Returns:
      gfapy.line.segment.GFA2 or str : instance or name of the other segment
        of the connection (the segment itself, if the connection is circular)
    """
    if isinstance(segment, gfapy.Line):
      segment = segment.name
    if segment == self.sid1.name:
      return self.sid2.line
    elif segment == self.sid2.name:
      return self.sid1.line
    else:
      raise gfapy.NotFoundError(
        "Line {} does not involve segment {}".format(self, segment))
